### PR TITLE
Shorten demos during testing outside of original scripts

### LINKED ISSUE #173: Run demos directly during testing
Some of the code coverage information is missing because we are running the demos using subprocesses.

Perhaps we could instead run them directly with Pytest and have the Pytest configuration delete any outputs.

### LINKED ISSUE #175: Use `sed` instead of `GOALIE_REGRESSION_TEST`
Rather than confusing the demo content with environment variables, we should apply specific modifications with `sed` when running the test suite.

### PR DESCRIPTION
**Pull request by ddundo**
Closes #173. Closes #175.

This PR:
- runs the demos directly in `test_demos.py` instead of using `subprocess`, so that demos contribute properly to the coverage report
- modifies the demos inside `test_demos.py` to cut down run time, instead of exporting a global environment variable that is then used inside demos in an `if` statement

@jwallwork23 you said that we can take over issues without asking, so I hope you don't mind that I did :)

### CODE REVIEW

**Comment by ddundo** [comment]
It looks like the tape isn't cleared properly, since I don't get these 3 failing tests if I just run `pytest test_mesh_seq.py`. This issue hasn't been caught because demos were run in a subprocess so it didn't affect other tests.

I will fix this later today/tomorrow.

File: test_adjoint/test_demos.py [hunk 1]
```diff
@@ -4,27 +4,70 @@
...
+# Modifications dictionary to cut down run time of demos:
+# - Each key is the name of a demo script to be modified
+# - Each value is a dictionary where:
+# -- The key is the original string or regex pattern that we wish to replace
+# -- The value is the replacement string (use "" to remove the original code)
+modifications = {
+    "burgers-hessian.py": {""""maxiter": 35""": """"maxiter": 3"""},
+    "gray_scott.py": {
+        "end_time = 2000.0": "end_time = 10.0",
+        r"ic = mesh_seq.get_initial_condition\(\)\n.*?adj_outfile.write\(\*adj_end.subfunctions\)": "",
```
Comment (jwallwork23) [inline test_adjoint/test_demos.py:28]: These bits are kind of ugly but hopefully we can clean them up when we address #117.

**Review by jwallwork23 — APPROVED** [review]
Nice! This is very cool, thanks.

File: test_adjoint/test_demos.py (hunk 1 above)
Comment (ddundo) [reply test_adjoint/test_demos.py:28]: Indeed, also hard to immediately see what's happening. I'll keep an eye on it and see if I can improve it :)

**Pull request merged**